litmus-portal/graphql-server/pkg/analytics/handler/handler.go

```diff
@@ -170,23 +211,225 @@ func UpdateDataSource(datasource model.DSInput) (*model.DSResponse, error) {
+	for _, applicationMetadata := range dashboard.ApplicationMetadataMap {
+		var newApplications []*dbSchemaAnalytics.Resource
+		for _, application := range applicationMetadata.Applications {
+			newApplication := dbSchemaAnalytics.Resource{
+				Kind:  application.Kind,
+				Names: application.Names,
+			}
+			newApplications = append(newApplications, &newApplication)
+		}
+		newApplicationMetadata := dbSchemaAnalytics.ApplicationMetadata{
+			Namespace:    applicationMetadata.Namespace,
+			Applications: newApplications,
+		}
+
+		newApplicationMetadataMap = append(newApplicationMetadataMap, newApplicationMetadata)
+	}
 
-	err := dbOperationsAnalytics.UpdateDashboard(query, update)
+	for i, panelGroup := range dashboard.PanelGroups {
+
+		var panelGroupID string
+
+		if panelGroup.PanelGroupID == "" {
+			panelGroupID = uuid.New().String()
+		} else {
+			panelGroupID = panelGroup.PanelGroupID
+			updatedDashboardPanelGroupIDs = append(updatedDashboardPanelGroupIDs, panelGroup.PanelGroupID)
+		}
+
+		newPanelGroups[i].PanelGroupID = panelGroupID
+		newPanelGroups[i].PanelGroupName = panelGroup.PanelGroupName
+
+		for _, panel := range panelGroup.Panels {
+
+			var (
+				panelID   string
+				createdAt string
+				updatedAt string
+			)
+
+			if *panel.PanelID == "" {
+				panelID = uuid.New().String()
+				createdAt = strconv.FormatInt(time.Now().Unix(), 10)
+				updatedAt = strconv.FormatInt(time.Now().Unix(), 10)
+			} else {
+				panelID = *panel.PanelID
+				createdAt = *panel.CreatedAt
+				updatedAt = strconv.FormatInt(time.Now().Unix(), 10)
+				updatedDashboardPanelIDs = append(updatedDashboardPanelIDs, *panel.PanelID)
+			}
+
+			var newPromQueries []*dbSchemaAnalytics.PromQuery
+			err := copier.Copy(&newPromQueries, &panel.PromQueries)
+			if err != nil {
+				return "error updating queries", err
+			}
+
+			var newPanelOptions dbSchemaAnalytics.PanelOption
+			err = copier.Copy(&newPanelOptions, &panel.PanelOptions)
+			if err != nil {
+				return "error updating options", err
+			}
+
+			newPanel := dbSchemaAnalytics.Panel{
+				PanelID:      panelID,
+				PanelOptions: &newPanelOptions,
+				PanelName:    panel.PanelName,
+				PanelGroupID: panelGroupID,
+				PromQueries:  newPromQueries,
+				IsRemoved:    false,
+				XAxisDown:    panel.XAxisDown,
+				YAxisLeft:    panel.YAxisLeft,
+				YAxisRight:   panel.YAxisRight,
+				Unit:         panel.Unit,
+				CreatedAt:    createdAt,
+				UpdatedAt:    updatedAt,
+			}
+
+			if *panel.PanelID == "" {
+				panelsToCreate = append(panelsToCreate, &newPanel)
+			} else {
+				panelsToUpdate = append(panelsToUpdate, &newPanel)
+			}
+		}
+	}
+
+	query := bson.D{
+		{"db_id", dashboard.DbID},
+		{"is_removed", false},
+	}
+
+	existingDashboard, err := dbOperationsAnalytics.GetDashboard(query)
+	if err != nil {
+		return "error fetching dashboard details", fmt.Errorf("error on query from dashboard collection by projectid: %v", err)
```

Comment:
yes but the problem then is to return an empty string since on success we are returning a text message it may be a good practice to return the sting and as for the custom part, the error returned may be a database error and we don't want the user to know the specifics of the database error over the graphQL query (it might be better to use a custom message which may be more understandable to the end user and also for logging purposes) its not a panic.

Reply:
We are returning two parameters: string and error, I'm talking about the error part. Instead of adding a custom error message, we can only return err, that we are doing now as well. 
I'm just suggesting to replace ` fmt.Errorf("error on query from dashboard collection by projectid: %v", err)` with `err`

Reply:
Yes as i mentioned that may spill the collection name etc if its a database error or something. custom message will let us control the logging (also sometimes error returned span a lot of lines and corrupt the server logs) it may be better this way cc: @rajdas98 @gdsoumya There had been a similar case where no data in collection user or something was being logged continuously sometime back.

Reply:
It's fine if you keep it this way, but you are printing err now as well.